### PR TITLE
ws-auth: Allow DNAME in apex with SOA and NS records

### LINKED ISSUE #8641: DNAME at apex via API leads to "Conflicts with other RRSet"
- Program: Authoritative
 - Issue type: Bug report

### Short description
Attempts to create a zone with a single DNAME via the API at the apex of the domain fail with the error: "Conflicts with another RRset"

### Environment
 - Operating system: Ubuntu (Bionic)
 - Software version: 4.2.1
 - Software source: deb [arch=amd64] http://repo.powerdns.com/ubuntu bionic-auth-42 main

### Steps to reproduce
1. Attempt to post `{"kind": "master", "name": "example.net.", "rrsets": [{"ttl": 3600, "name": "example.net.", "records": [{"content": "example.co.uk.", "disabled": false}], "type": "DNAME"}], "masters": [], "nameservers": []}` to the Zones API at `http://localhost:8081/api/v1/servers/localhost/zones`
2. Observe a reply `{"error": "RRset example.net. IN DNAME: Conflicts with another RRset"}`

### Expected behaviour
I'd expect the zone to be created, with a single DNAME at the apex of the domain.

### Actual behaviour
422 error from the API citing that it conflicts with another RRset.  No other RRset exists.

### Other information
I discussed this briefly on IRC with dwfreed - who assured me that DNAMEs could exist with other records.  I think, however, there's something odd going on with the apex of a domain here.

While I vaguely understand DNAMEs, I'm unsure if DNAME at the apex is permissible - seemingly not in PDNS as it stands, but I think it is in BIND (which one of my customers is moving away from).

### PR DESCRIPTION
**Pull request by cmouse**
### Short description
Permits DNAME with SOA and NS in zone apex. Closes #8641.

I have:
- [x] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [x] compiled this code
- [ ] tested this code
- [ ] included documentation (including possible behaviour changes)
- [ ] documented the code
- [x] added or modified regression test(s)
- [ ] added or modified unit test(s)

### CODE REVIEW

File: pdns/ws-auth.cc [hunk 1]
```diff
@@ -1356,11 +1357,31 @@ static void checkNewRecords(vector<DNSResourceRecord>& records) {
...
       }
     }
 
+    // check apex records
+    if (apexRecords.count(QType::CNAME) > 0)
+       throw ApiException("Apex record has CNAME");
+    if (apexRecords.count(QType::DNAME) > 0 && apexRecords[QType::DNAME] > 1)
+       throw ApiException("Apex record has duplicate DNAME");
+    if (apexRecords.count(QType::SOA) > 0 && apexRecords[QType::SOA] > 1)
+       throw ApiException("Apex record has duplicate DNAME");
```
Comment (Habbie) [inline pdns/ws-auth.cc]:
```suggestion
       throw ApiException("Apex record has duplicate SOA");
```

File: pdns/ws-auth.cc [hunk 2]
```diff
@@ -1356,11 +1357,31 @@ static void checkNewRecords(vector<DNSResourceRecord>& records) {
...
+    if (apexRecords.count(QType::DNAME) > 0) {
+       // allow SOA and NS
+       for(auto it = apexRecords.cbegin(); it != apexRecords.cend(); it++)
+         if (it->first != QType::DNAME && it->first != QType::SOA && it->first != QType::NS)
+            throw ApiException("Apex with DNAME can only hae SOA and NS records");
```
Comment (Habbie) [inline pdns/ws-auth.cc]: This is wrong. A, MX, TXT, etc. can all coexist with DNAME.

Comment (Habbie) [reply pdns/ws-auth.cc]: And this is not an apex rule, this is a general rule.

File: regression-tests.api/test_Zones.py [hunk 3]
```diff
@@ -1298,31 +1298,6 @@ def test_rrset_unknown_type(self):
 
     @parameterized.expand([
         ('CNAME', ),
-        ('DNAME', ),
-    ])
-    def test_rrset_exclusive_and_other(self, qtype):
```
Comment (zeha) [inline regression-tests.api/test_Zones.py]: this test is still valid for CNAMEs, no?

**Comment by cjsoftuk** [comment]
What can I reasonably do to help push this forward?  Do you need me to test it in the specific case that triggered the original report?

**Comment by Habbie** [comment]
> Do you need me to test it in the specific case that triggered the original report?

That never hurts!

**Comment by cjsoftuk** [comment]
This seems to have fixed one of the issues that was preventing us from moving to PDNS - so thank you.  Given that it seems to be OK - is there anything else productive I can do to help this down the road?

**Comment by Habbie** [comment]
You've done what you could! This should be reviewed and merged soon.

File: pdns/ws-auth.cc [hunk 4]
```diff
@@ -2029,23 +2036,37 @@ static void patchZone(UeberBackend& B, HttpRequest* req, HttpResponse* resp) {
...
+          getAllRecords(di.backend, records);
+
+          for(auto& existing_rrset: records) {
+            bool dname_seen = false, ns_seen = false;
+            sort(existing_rrset.second.begin(), existing_rrset.second.end(),
+              [](const DNSResourceRecord& rec_a, const DNSResourceRecord& rec_b) -> bool {
+              /* we need _strict_ weak ordering */
+                return std::tie(rec_a.qname, rec_a.qtype, rec_a.content) < std::tie(rec_b.qname, rec_b.qtype, rec_b.content);
+              }
+            );
```
Comment (mind04) [inline pdns/ws-auth.cc]: Please explain why you need the sorting. Lookup will only return records from one qname, and the code below is  iterating over all rr's.

File: pdns/ws-auth.cc [hunk 5]
```diff
@@ -2029,23 +2036,37 @@ static void patchZone(UeberBackend& B, HttpRequest* req, HttpResponse* resp) {
...
+            for(const auto& rr: existing_rrset.second) {
```
Comment (mind04) [inline pdns/ws-auth.cc]: Without the sorting a normal get() will do in this loop and will avoid some copying of rrs.

Comment (cmouse) [reply pdns/ws-auth.cc]: We agreed to add the getAllRecords here so it might eventually maybe replace get().

File: pdns/ws-auth.cc (hunk 4 above)
Comment (cmouse) [reply pdns/ws-auth.cc]: I *suppose* we don't need to sort them for this.

**Comment by cjsoftuk** [comment]
Hi guys - any news on this?  Sorry to keep chasing!

**Pull request merged**